File: Core/src/main/java/com/github/ezauton/core/action/TimedPeriodicAction.java
package com.github.ezauton.core.action;

import java.util.concurrent.TimeUnit;

/**
 * Describes an action that ends after a certain amount of time has elapsed
 */
public class TimedPeriodicAction extends PeriodicAction {

    private long durationMillis;

    /**
     * @param period       How long to (repeatedly) run the runnables for
     * @param periodUnit
     * @param duration     The timeunit that period is
     * @param durationUnit
     */
    public TimedPeriodicAction(long period, TimeUnit periodUnit, long duration, TimeUnit durationUnit, Runnable... runnables) {
        super(period, periodUnit, runnables);
        durationMillis = durationUnit.toMillis(duration);
    }

    /**
     * @param period       How long to (repeatedly) run the runnables for
     * @param periodUnit
     * @param duration     The timeunit that period is
     * @param durationUnit
     */
    public TimedPeriodicAction(long period, TimeUnit periodUnit, long duration, TimeUnit durationUnit) {
        super(period, periodUnit);
        durationMillis = durationUnit.toMillis(duration);
    }

    /**
     * @param duration     The timeunit that period is
     * @param durationUnit
     */
    public TimedPeriodicAction(long duration, TimeUnit durationUnit, Runnable... runnables) {
        super(runnables);
        durationMillis = durationUnit.toMillis(duration);
    }

    /**
     * @param duration     The timeunit that period is
     * @param durationUnit
     */
    public TimedPeriodicAction(long duration, TimeUnit durationUnit) {
        super();
        durationMillis = durationUnit.toMillis(duration);
    }

    @Override
    protected boolean isFinished() {
        return this.stopwatch != null && this.stopwatch.read() > durationMillis;
    }

}

File: Core/src/main/java/com/github/ezauton/core/action/TimedPeriodicAction.kt
package com.github.ezauton.core.action

import com.github.ezauton.core.Duration
import java.util.concurrent.TimeUnit

/**
 * Describes an action that ends after a certain amount of time has elapsed
 */
class TimedPeriodicAction : PeriodicAction {

    private var durationMillis: Long = 0

    /**
     * @param period       How long to (repeatedly) run the runnables for
     * @param periodUnit
     * @param duration     The timeunit that period is
     * @param durationUnit
     */
    constructor(period: Duration = DEFAULT_PERIOD, duration: Duration , vararg runnables: Runnable) : super(period, *runnables) {
        durationMillis = duration.millis
    }

    /**
     * @param duration     The timeunit that period is
     * @param durationUnit
     */
    constructor(duration: Duration, vararg runnables: Runnable) : super(*runnables) {
        durationMillis = duration.millis
    }

    /**
     * @param duration     The timeunit that period is
     * @param durationUnit
     */
    constructor(duration: Long, durationUnit: TimeUnit) : super() {
        durationMillis = durationUnit.toMillis(duration)
    }

    override fun isFinished(): Boolean {
        return this.stopwatch.read() > durationMillis
    }

}
